lib/std/array_list.zig

```diff
@@ -816,9 +803,45 @@ pub fn ArrayListAlignedUnmanaged(comptime T: type, comptime alignment: ?u29) typ
             len: usize,
             new_items: []const T,
         ) Allocator.Error!void {
-            var managed = self.toManaged(allocator);
-            defer self.* = managed.moveToUnmanaged();
-            try managed.replaceRange(start, len, new_items);
+            const after_range = try addOrOom(start, len);
+            const range = self.items[start..after_range];
+            if (range.len < new_items.len) {
+                const first = new_items[0..range.len];
+                const rest = new_items[range.len..];
+                @memcpy(range[0..first.len], first);
+                try self.insertSlice(allocator, after_range, rest);
+            } else {
+                self.replaceRangeAssumeCapacity(start, len, new_items);
+            }
+        }
+
+        /// Replace range of elements `list[start..][0..len]` with `new_items`.
+        /// Grows list if `len < new_items.len`.
+        /// Shrinks list if `len > new_items.len`.
+        /// Invalidates pointers if this ArrayList is resized.
+        pub fn replaceRangeAssumeCapacity(self: *Self, start: usize, len: usize, new_items: []const T) void {
+            const after_range = try addOrOom(start, len);
```

Comment:
@andrewrk I had to manually compare the diffs because the code moved. I believe wrapping this in addOrOom is the only substantive change upstream (and same in non-assume version of the function). You had also added that in the shrinking list case but I refactored the addition to a subtraction. I believe it is safe but please have a look.

Reply:
`addOrOom` is only needed (and only will compile successfully) when used in a function that possibly allocates. In an "assume capacity" function such as this one, it returns void because it cannot fail. No need to call `addOrOom` because the caller is asserting that there is enough capacity already. In other words, it can continue to use `+` just like before.

Reply:
Ah ok that makes sense. But then I am effectively removing this change:

https://github.com/ziglang/zig/commit/69461bcae4a78c835bdfe0aae85524342c0f8461#diff-94a89ceea2dcbee151b0120728ddcfc809f0c2b9c66a545547b4d25271cb1b6aR247


Is that a problem?

Reply:
replaceRange (fallible) should call addOrOom(). replaceRangeAssumeCapacity should not

Reply:
But then I cannot do the `moveToManaged` shortcut right? I have to duplicate all the logic so I can switch to `addOrOom`.

Reply:
no, you can use the moveToUnmanaged strategy in both of the functions of ArrayListManaged

Reply:
Ah okay sorry I get it now - will update. Going to rebuild so I can run the tests again too.